File: python/questions/largestRectangleArea.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        lh = len(heights)
        st = [-1]
        ans = 0
        for i, h in enumerate(heights):
            while len(st)>1 and h<=heights[st[-1]]:
                ph = heights[st.pop()]
                ans = max(ans, ph*(i-st[-1]-1))
            st.append(i)
        while len(st)>1:
            ph = heights[st.pop()]
            ans = max(ans, ph*(lh-st[-1]-1))
        return ans

File: cpp/questions/largestRectangleArea.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int hs = heights.size();
        int ans = 0;
        std::stack<int> st;
        st.push(-1);
        for (int p=0; p<hs; ++p) {
            int ch = heights[p];
            while (st.size()>1 && ch<=heights[st.top()]) {
                int ph = heights[st.top()];
                st.pop();
                ans = std::max(ans, ph*(p-st.top()-1));
            }
            st.push(p);
        }
        while (st.size() > 1) {
            int ph = heights[st.top()];
            st.pop();
            ans = std::max(ans, ph*(hs-st.top()-1));
        }
        return ans;
    }
};
